File: csharp/UIGeneration/TabbedControlPanelOrganizer.cs
using System;
using System.Collections;
using PUC;

namespace PUC.UIGeneration
{
	/// <summary>
	/// This organizer creates a set of panels that are controlled by a
	/// widget outside of the panels.  It is assumed that such a widget 
	/// exists (and is typically enforced by some manipulation to the group
	/// earlier in the generation process)
	/// </summary>
	public class TabbedControlPanelOrganizer : ControlOrganizer
	{
		/*
		 * Member Variables
		 */
	
		protected UIGenerator _ui;
		protected bool _vertical;


		/*
		 * Constructor
		 */

		public TabbedControlPanelOrganizer( UIGenerator ui,
			ApplianceState state,
			ArrayList children,
			ArrayList dependencies,
			bool vertical
			)
			: base( state )
		{
			_ui = ui;
			_vertical = vertical;

			_dependencies = new Hashtable( children.Count );

			for( int i = 0; i < children.Count; i++ )
			{
				GroupNode g = (GroupNode)children[ i ];
				ArrayList deps = (ArrayList)dependencies[ i ];

				_dependencies[ g ] = deps;
			}
		}


		/*
		 * Organizer Method
		 */

		public override Hashtable AddOrganization(GroupNode group, InterfaceNode currentNode)
		{
			TabbedOverlappingPanelsNode tab = new TabbedOverlappingPanelsNode( _state );
			MultiplePanelNode non = null;

			if ( _vertical )
				non = new HorizontallyNonOverlappingPanelsNode( (PanelNode)currentNode, tab );
			else
			{
				non = new VerticallyNonOverlappingPanelsNode();
				currentNode.InsertAsParent( non );
				non.AddPanel( tab );
			}

			Hashtable deps = new Hashtable( _dependencies.Count );
			Hashtable panels = new Hashtable( _dependencies.Count );

			IEnumerator e = _dependencies.Keys.GetEnumerator();
			PanelNode newP = null;
			while( e.MoveNext() )
			{
				GroupNode g = (GroupNode)e.Current;
				ArrayList aryDeps = (ArrayList)_dependencies[ g ];

				EqualsDependency eqDep = (EqualsDependency)aryDeps[ 0 ];

				try
				{
					newP = (PanelNode)tab.GetNodeByValue( (int)eqDep.Value ).GetChildNode();
					newP.Group = g;
				}
				catch( Exception ) { Globals.GetFrame( eqDep.State.Appliance ).AddLogLine( "Error in TabbedControlPanelOrganizer... Line 82" ); }

				deps[ aryDeps ] = newP;
				panels[ g ] = newP;
			}

			tab.InitGroupToTabMap();

			_activePanel = newP;

			_dependencies = deps;
			_uiValid = true;

			this.ValueChanged( _state );

			return panels;
		}
	}
}

File: edu/cmu/hcii/puc/uigen/TabbedControlPanelOrganizer.java
/**
 * TabbedControlPanelOrganizer.java
 *
 * This organizer creates a set of panels that are controlled by a
 * widget outside of the panels.  It is assumed that such a widget
 * exists (and is typically enforced by some manipulation to the group
 * tree earlier in the generation process) 
 */

// Package Definition

package edu.cmu.hcii.puc.uigen;


// Import Declarations

import java.util.Enumeration;
import java.util.Hashtable;
import java.util.Vector;

import edu.cmu.hcii.puc.ApplianceState;
import edu.cmu.hcii.puc.EqualsDependency;
import edu.cmu.hcii.puc.GroupNode;


// Class Definition

public class TabbedControlPanelOrganizer extends ControlOrganizer {

    //**************************
    // Member Variables
    //**************************

    protected boolean m_bVertical;


    //**************************
    // Constructor
    //**************************

    public TabbedControlPanelOrganizer( UIGenerator ui,
					ApplianceState pState,
					Vector vChildren,
					Vector vDependencies,
					boolean bVertical
					) {
	super( ui, pState );

	m_bVertical = bVertical;

	m_hDependencies = new Hashtable( vChildren.size() );

	for( int i = 0; i < vChildren.size(); i++ ) {
	    GroupNode g = (GroupNode)vChildren.elementAt( i );
	    Vector vDeps = (Vector)vDependencies.elementAt( i );

	    m_hDependencies.put( g, vDeps );
	}
    }


    //**************************
    // Organizer Method
    //**************************

    public Hashtable addOrganization( GroupNode pGroup, 
				      InterfaceNode pCurrentNode ) {

	TabbedOverlappingPanelsNode pTab = new TabbedOverlappingPanelsNode( m_pUI.m_pAppliance, m_pState );
	MultiplePanelNode pNon;
	if ( m_bVertical ) 
	    pNon = new VerticalNoOverlapPanelsNode( (PanelNode)pCurrentNode,
						    pTab );
	else {
	    pNon = new NonOverlappingPanelsNode();
	    pCurrentNode.insertAsParent( pNon );
	    pNon.addPanel( pTab );
	}

	Hashtable hDeps = new Hashtable( m_hDependencies.size() );
	Hashtable hPanels = new Hashtable( m_hDependencies.size() );

	Enumeration en = m_hDependencies.keys();
	PanelNode newP = null;
	while( en.hasMoreElements() ) {
	    GroupNode g = (GroupNode)en.nextElement();
	    Vector vDeps = (Vector)m_hDependencies.get( g );

	    EqualsDependency pDep = (EqualsDependency)vDeps.elementAt( 0 );

	    try {
		// assuming that getChildNode() returns an
		// InterfaceNode that is a PanelNode
		newP = (PanelNode)pTab.getNodeByValue( new Integer( pDep.getValueString() ).intValue() ).getChildNode();
	    }
	    catch( Throwable t ) { t.printStackTrace(); }

	    hDeps.put( vDeps, newP );
	    hPanels.put( g, newP );
	}

	m_pActivePanel = newP;

	m_hDependencies = hDeps;
	m_bUIValid = true;

	this.valueChanged( m_pState );

	return hPanels;
    }
}
